marlowe-runtime/src/Language/Marlowe/Runtime/Transaction/BuildConstraints.hs

```diff
@@ -396,11 +423,34 @@ buildApplyInputsConstraintsV1 systemStart eraHistory marloweOutput invalidBefore
 
     nextMarloweTimeout :: V1.Contract -> Maybe UTCTime
     nextMarloweTimeout (V1.When _ timeout _) = Just $ posixTimeToUTCTime timeout
-    nextMarloweTimeout _ = Nothing
-
+    nextMarloweTimeout V1.Close = Nothing
+    nextMarloweTimeout (V1.Pay _ _ _ _ c) = nextMarloweTimeout c
+    nextMarloweTimeout (V1.If _ c1 c2) = on min nextMarloweTimeout c1 c2
+    nextMarloweTimeout (V1.Let _ _ c) = nextMarloweTimeout c
+    nextMarloweTimeout (V1.Assert _ c) = nextMarloweTimeout c
+
+    maxSafeSlot :: O.SlotNo
+    maxSafeSlot = getMaxSafeSlotFromSummary $ unInterpreter interpreter
+
+    unInterpreter :: Interpreter xs -> Summary xs
+    unInterpreter = unsafeCoerce -- Interpreter xs is a newtype wrapper around Summary xs.
+                                 -- Unfortunately, the provided query language is not able to extract the max safe slot,
+                                 -- So we have to reach unsafely into the internals to get
```

Comment:
Should we submit a ticket to the ouroboros repo?

Reply:
Yes, perhaps that would be a good idea.